ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs

```diff
@@ -636,13 +644,147 @@ data LedgerBackingStoreValueHandle m l = LedgerBackingStoreValueHandle
+              rew   = rewindTableKeySets chlog ks
+          unfwd <- readKeySetsVH
+                     (\ks' -> (,) seqNo <$> BackingStore.bsvhRead vh ks')
+                     rew
+          case forwardTableKeySets chlog unfwd of
+              Left _err -> error "impossible!"
+              Right vs  -> pure vs
+      )
+      (\rq -> do
+          let chlog = ledgerDbChangelog ldb
+              -- Get the differences without the keys that are greater or equal
+              -- than the maximum previously seen key.
+              diffs =
+                maybe
+                  id
+                  (zipLedgerTables doDropLTE)
+                  (BackingStore.rqPrev rq)
+                  $ mapLedgerTables prj
+                  $ changelogDiffs chlog
+              -- (1) Ensure that we never delete everything read from disk (ie
+              --     if our result is non-empty then it contains something read
+              --     from disk).
+              --
+              -- (2) Also, read one additional key, which we will not include in
+              --     the result but need in order to know which in-memory
+              --     insertions to include.
+              maxDeletes = maybe 0 getMax
+                         $ foldLedgerTables (Just . Max . numDeletesDiffMK) diffs
+              nrequested = 1 + max (BackingStore.rqCount rq) (1 + maxDeletes)
+
+          values <- BackingStore.bsvhRangeRead vh (rq{BackingStore.rqCount = nrequested})
+          pure $ zipLedgerTables (doFixupReadResult nrequested) diffs values
+      )
+      close
+  where
+    prj ::
+         Ord k
+      => ApplyMapKind SeqDiffMK k v
+      -> ApplyMapKind DiffMK k v
+    prj (ApplySeqDiffMK sq) = ApplyDiffMK (HD.cumulativeDiffSeqUtxoDiff sq)
+
+    -- remove all diff elements that are <= to the greatest given key
+    doDropLTE ::
+         Ord k
+      => ApplyMapKind KeysMK k v
+      -> ApplyMapKind DiffMK k v
+      -> ApplyMapKind DiffMK k v
+    doDropLTE (ApplyKeysMK (HD.UtxoKeys ks)) (ApplyDiffMK (HD.UtxoDiff ds)) =
+        ApplyDiffMK
+      $ HD.UtxoDiff
+      $ case Set.lookupMax ks of
+          Nothing -> ds
+          Just k  -> Map.filterWithKey (\dk _dv -> dk > k) ds
+
+    -- NOTE: this is counting the deletions wrt disk.
+    numDeletesDiffMK :: ApplyMapKind DiffMK k v -> Int
+    numDeletesDiffMK (ApplyDiffMK (HD.UtxoDiff m)) =
+      getSum $ foldMap (Sum . oneIfDel) m
+      where
+        oneIfDel (HD.UtxoEntryDiff _v diffstate) = case diffstate of
+          HD.UedsDel       -> 1
+          HD.UedsIns       -> 0
+          HD.UedsInsAndDel -> 0
+
+    -- INVARIANT: nrequested > 0
+    --
+    -- (1) if we reached the end of the store, then simply yield the given diff
+    --     applied to the given values
+    -- (2) otherwise, the readset must be non-empty, since 'rqCount' is positive
+    -- (3) remove the greatest read key
+    -- (4) remove all diff elements that are >= the greatest read key
```

Comment:
If I'm reading the code correctly, I think it is doing the opposite of (4)

Reply:
Are you referring to this line? 

```
HD.UtxoDiff $ Map.filterWithKey (\dk _dv -> dk < k) ds
```

Here we're filtering *out* all the differences whose key is >= the greatest read key (`k` in this case).